File: DSA/saumya/Day15_Q2.py
class Node:
  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None

def level_order_traversal(root):
  if not root:
    return []

  traversal = []
  queue = [root]
  while queue:
    level_size = len(queue)
    for _ in range(level_size):
      current_node = queue.pop(0)
      traversal.append(current_node.data)
      if current_node.left:
        queue.append(current_node.left)
      if current_node.right:
        queue.append(current_node.right)
  return traversal

# test case
tree = Node(3)
tree.left = Node(9)
tree.right = Node(20)
tree.right.left = Node(15)
tree.right.right = Node(7)

level_order = level_order_traversal(tree)
print("Level order traversal:", level_order)

File: DSA/Vyom-Singhal/Day15/Day15-q2.cpp
#include <iostream>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

void levelOrderTraversal(TreeNode* root) {
    if (root == NULL) return;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        TreeNode* current = q.front();
        q.pop();
        cout << current->val << " ";
        if (current->left != NULL) q.push(current->left);
        if (current->right != NULL) q.push(current->right);
    }
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    cout << "Level Order Traversal: ";
    levelOrderTraversal(root);
    delete root->left->left;
    delete root->left->right;
    delete root->left;
    delete root->right;
    delete root;
}
